Fixes #27039 - Content View version Export Condition Fix

patilsuraj767 opened the pull request:
"repositories" return the empty array if there is no repo added in the content view. So we should check the length of the array instead checking if it is set.

akofink commented on lib/hammer_cli_katello/content_view_version.rb:
```diff
@@ -335,6 +335,8 @@ def create_tar(cv, cvv, repositories, json)
             `tar cvfh #{repo_tar} #{repo_dirs.join(" ")}`
           end
+        else
+          puts "Please verify that content view version id - #{cvv['id']} contents atleast one repository"
```
s/atleast/at least/

chris1984 replied:
s/contents/contains

akofink commented on lib/hammer_cli_katello/content_view_version.rb:
```diff
@@ -324,7 +324,7 @@ def create_tar(cv, cvv, repositories, json)
         Dir.mkdir("#{export_dir}/#{export_prefix}")
 
-        if repositories
+        if repositories.length > 0
```
`if repositories.any?`

patilsuraj767 commented:
@chris1984 @akofink Sorry I have checked the you update after pushing the commit. 

I have modified the whole message as it was going beyond 100 char and test was failing. 

Also changed   if repositories.length > 0  To   if !repositories.empty? let me if that is fine.

akofink commented on lib/hammer_cli_katello/content_view_version.rb:
```diff
@@ -324,7 +324,7 @@ def create_tar(cv, cvv, repositories, json)
         Dir.mkdir("#{export_dir}/#{export_prefix}")
 
-        if repositories
+        if !repositories.empty?
```
This can be `if repositories.any?`

patilsuraj767 replied:
Ok I will change it. Thanks.

akofink commented on lib/hammer_cli_katello/content_view_version.rb:
```diff
@@ -335,6 +335,8 @@ def create_tar(cv, cvv, repositories, json)
             `tar cvfh #{repo_tar} #{repo_dirs.join(" ")}`
           end
+        else
+          puts "Ensure the content view '#{cvv['name']}' has atleast one repository."
```
There's still a typo here: s/atleast/at least/

patilsuraj767 replied:
Yes, will change this too.

chris1984 commented:
@patilsuraj767 I will start testing this tomorrow and work on letting you know why the tests are failing. Assigning to me

patilsuraj767 commented:
@chris1984 Sure, Thanks.

akofink commented on lib/hammer_cli_katello/content_view_version.rb:
```diff
@@ -324,7 +324,7 @@ def create_tar(cv, cvv, repositories, json)
         Dir.mkdir("#{export_dir}/#{export_prefix}")
 
-        if repositories
+        if repositories.any?
```
`s/repositories.any?/repositories&.any?/` will fix your tests

patilsuraj767 replied:
@akofink Thanks that worked

akofink commented on lib/hammer_cli_katello/content_view_version.rb:
```diff
@@ -335,6 +335,8 @@ def create_tar(cv, cvv, repositories, json)
             `tar cvfh #{repo_tar} #{repo_dirs.join(" ")}`
           end
+        else
+          puts "Ensure the content view '#{cvv['name']}' has at least one repository."
```
Hmm... I just realized another issue. The test that was failing was for composite content view export (in which repositories is nil). This message doesn't make sense in that context.

akofink replied:
IMO you can fix the composite issue by checking `elsif !cv['composite']`

patilsuraj767 commented:
I think we will need to put the below two "Dir" blocks inside the if condition, because it always create the .tar file, even if the command is failing with the error.
Let me know your thoughts on this @chris1984 @akofink  

```
       Dir.chdir("#{export_dir}/#{export_prefix}") do
          File.open(export_file, 'w') do |file|
            file.write(JSON.pretty_generate(json))
          end
        end

        Dir.chdir(export_dir) do
          `tar cf #{export_tar} #{export_prefix}`
          FileUtils.rm_rf(export_prefix)
        end

```

chris1984 commented:
> I think we will need to put the below two "Dir" blocks inside the if condition, because it always create the .tar file, even if the command is failing with the error.
> Let me know your thoughts on this @chris1984 @akofink
> 
> ```
>        Dir.chdir("#{export_dir}/#{export_prefix}") do
>           File.open(export_file, 'w') do |file|
>             file.write(JSON.pretty_generate(json))
>           end
>         end
> 
>         Dir.chdir(export_dir) do
>           `tar cf #{export_tar} #{export_prefix}`
>           FileUtils.rm_rf(export_prefix)
>         end
> ```

If we do that the tests will prob break and need updating but yeah imho creating a tar with nothing in on error is a bad experience for the user

chris1984 commented:
Thanks @patilsuraj767 ill test again, @akofink does this look good? If so ill merge after testing

akofink commented on lib/hammer_cli_katello/content_view_version.rb:
```diff
@@ -335,17 +334,22 @@ def create_tar(cv, cvv, repositories, json)
+          end
+
+        elsif cv['composite']
+          puts "Cannot export composite content view."
```
Definitely not true - composite CVs are exported without repositories, just the json.

akofink commented on lib/hammer_cli_katello/content_view_version.rb:
```diff
@@ -322,9 +322,8 @@ def create_tar(cv, cvv, repositories, json)
 
-        Dir.mkdir("#{export_dir}/#{export_prefix}")
-
-        if repositories
```
You need to check if the content view is not composite here, and completely skip this block if it is.

patilsuraj767 commented:
@akofink My bad. I thought we cannot export the content view as there is no repository in it.
I have changed the code again and added the test cases back. As far as I have tested, now we are able to export the composite content view and also, there is no tar file created if the content view version export fails. Please review it again.

akofink commented on test/functional/content_view/version/export_test.rb:
```diff
@@ -90,7 +90,7 @@
     File.expects(:exist?).with('/usr/share/foreman').returns(true)
     File.stubs(:exist?).with('/var/log/hammer/hammer.log._copy_').returns(false)
-
+    
```
Please remove the vestigial white space changes.

akofink commented on lib/hammer_cli_katello/content_view_version.rb:
```diff
@@ -311,7 +311,11 @@ def execute
 
         json = export_json(export_json_options)
-        create_tar(cv, cvv, repositories, json)
+        if repositories&.any? or cv['composite']
```
Use `||` instead of `or`, due to [precedent issues](http://ruby-doc.org/core-2.6.3/doc/syntax/precedence_rdoc.html). I'm surprised rubocop didn't complain about this.

patilsuraj767 replied:
rubocop has complained about it. I just thought to correct it later, to avoid unnecessary commit.

akofink commented on lib/hammer_cli_katello/content_view_version.rb:
```diff
@@ -325,6 +329,7 @@ def create_tar(cv, cvv, repositories, json)
         Dir.mkdir("#{export_dir}/#{export_prefix}")
 
         if repositories
```
I think you still need to check if `repositories&.any?` here, as an empty repositories array (`[]`) will still be exported.

patilsuraj767 replied:
I think at this point empty repositories array will not be present because we are checking for empty repositories in the previous function. 

```
if repositories&.any? or cv['composite']
     create_tar(cv, cvv, repositories, json)
else
     puts "Ensure the content view version '#{cvv['name']}' has at least one repository."
end
```

akofink replied:
Could repositories still be `[]` though if it's a composite CV?

patilsuraj767 replied:
@akofink I do not think that repository can be `[]` in case of composite CV. 
Repository return `[]` only if any content view is published without adding any repository in it. 
In can of composite CV, I have observed that repository variable itself is not set. But anyways its good change it to `repositories&.any?`. I will make the changes and send the commit again.

patilsuraj767 commented:
@akofink Added the test case.

chris1984 commented:
Will retest and if it works ill merge. The code looks good :+1:
